R/exercise.R

```diff
@@ -1311,6 +1328,154 @@ local_restore_options_and_envvars <- function(.local_envir = parent.frame()) {
   local_restore_envvars(.local_envir)
 }
 
+# Render Exercise RMD -----------------------------------------------------
+# — Prep ----
+render_exercise_rmd_prep <- function(exercise, ...) {
+  UseMethod("render_exercise_rmd_prep", exercise)
+}
+
+#' @export
+render_exercise_rmd_prep.default <- function(exercise, ...) {
+  exercise_code_chunks_prep(exercise)
+}
+
+# — User ----
+render_exercise_rmd_user <- function(exercise, ...) {
+  UseMethod("render_exercise_rmd_user", exercise)
+}
+
+#' @export
+render_exercise_rmd_user.default <- function(exercise, ...) {
+  c(
+    readLines(system.file("internals", "templates", "exercise-setup.Rmd", package = "learnr")),
+    "",
+    exercise_code_chunks_user(exercise)
+  )
+}
+
+#' @export
+render_exercise_rmd_user.sql <- function(exercise, ...) {
+  rmd_src_user <- NextMethod()
+
+  c(
+    rmd_src_user,
+    "",
+    # knitr's sql chunk engine will either display the results or return the
+    # results back to R. We want both, so we ask knitr to return the result and
+    # then we explicitly print it in the chunk below.
+    '```{r eval=exists("___sql_result")}',
+    'get("___sql_result")',
+    "```"
+  )
+}
+
+#' @export
+render_exercise_rmd_user.python <- function(exercise, ...) {
+  rmd_src_user <- NextMethod()
+
+  c(
+    rmd_src_user,
+    "",
+    # this is how we get the `last_value` from the python session
+    '```{r include=FALSE}',
+    'reticulate::py_run_string("import builtins")',
+    'reticulate::py_eval("builtins._")',
```

Comment:
We will want to make sure to preserve the last value as a Python object:

```suggestion
    'reticulate::py_eval("builtins._", convert=FALSE)',
```